File: src/weaponsprocurement/core/WeaponStockSnapshot.java
package weaponsprocurement.core;

import com.fs.starfarer.api.campaign.econ.MarketAPI;

import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

public final class WeaponStockSnapshot {
    private final MarketAPI market;
    private final OwnedSourcePolicy ownedSourcePolicy;
    private final StockSortMode sortMode;
    private final boolean includeBlackMarket;
    private final StockSourceMode sourceMode;
    private final Map<StockCategory, List<WeaponStockRecord>> recordsByCategory;
    private final Map<StockItemType, Map<StockCategory, List<WeaponStockRecord>>> recordsByTypeAndCategory;
    private final Map<String, WeaponStockRecord> recordsByWeaponId;
    private final List<WeaponStockRecord> allRecords;
    private final int totalRecords;

    public WeaponStockSnapshot(MarketAPI market,
                               OwnedSourcePolicy ownedSourcePolicy,
                               StockSortMode sortMode,
                               boolean includeBlackMarket,
                               StockSourceMode sourceMode,
                               Map<StockCategory, List<WeaponStockRecord>> recordsByCategory) {
        this.market = market;
        this.ownedSourcePolicy = ownedSourcePolicy;
        this.sortMode = sortMode;
        this.includeBlackMarket = includeBlackMarket;
        this.sourceMode = sourceMode == null ? StockSourceMode.LOCAL : sourceMode;
        this.recordsByCategory = immutableCategoryMap(recordsByCategory);
        this.recordsByTypeAndCategory = immutableTypeCategoryMap(this.recordsByCategory);
        this.recordsByWeaponId = immutableWeaponMap(this.recordsByCategory);
        this.allRecords = immutableAllRecords(this.recordsByCategory);
        this.totalRecords = allRecords.size();
    }

    public MarketAPI getMarket() {
        return market;
    }

    public OwnedSourcePolicy getOwnedSourcePolicy() {
        return ownedSourcePolicy;
    }

    public StockSortMode getSortMode() {
        return sortMode;
    }

    public boolean isIncludeBlackMarket() {
        return includeBlackMarket;
    }

    public StockSourceMode getSourceMode() {
        return sourceMode;
    }

    public List<WeaponStockRecord> getRecords(StockCategory category) {
        List<WeaponStockRecord> records = recordsByCategory.get(category);
        return records == null ? Collections.<WeaponStockRecord>emptyList() : records;
    }

    public List<WeaponStockRecord> getRecords(StockItemType itemType, StockCategory category) {
        Map<StockCategory, List<WeaponStockRecord>> byCategory = recordsByTypeAndCategory.get(itemType);
        if (byCategory == null) {
            return Collections.emptyList();
        }
        List<WeaponStockRecord> records = byCategory.get(category);
        return records == null ? Collections.<WeaponStockRecord>emptyList() : records;
    }

    public int getCount(StockItemType itemType) {
        int count = 0;
        Map<StockCategory, List<WeaponStockRecord>> byCategory = recordsByTypeAndCategory.get(itemType);
        if (byCategory == null) {
            return 0;
        }
        for (StockCategory category : StockCategory.values()) {
            List<WeaponStockRecord> records = byCategory.get(category);
            count += records == null ? 0 : records.size();
        }
        return count;
    }

    public int getCount(StockCategory category) {
        return getRecords(category).size();
    }

    public int getTotalRecords() {
        return totalRecords;
    }

    public WeaponStockRecord getRecord(String weaponId) {
        if (weaponId == null) {
            return null;
        }
        return recordsByWeaponId.get(weaponId);
    }

    public List<WeaponStockRecord> getAllRecords() {
        return allRecords;
    }

    public String getMarketName() {
        if (StockSourceMode.FIXERS.equals(sourceMode)) {
            return GlobalWeaponMarketService.FIXERS_MARKET_NAME;
        }
        if (StockSourceMode.SECTOR.equals(sourceMode)) {
            return GlobalWeaponMarketService.SECTOR_MARKET_NAME;
        }
        return market == null ? "No market context" : market.getName();
    }

    private static Map<StockCategory, List<WeaponStockRecord>> immutableCategoryMap(Map<StockCategory, List<WeaponStockRecord>> source) {
        Map<StockCategory, List<WeaponStockRecord>> result = new EnumMap<StockCategory, List<WeaponStockRecord>>(StockCategory.class);
        for (StockCategory category : StockCategory.values()) {
            List<WeaponStockRecord> records = source.get(category);
            if (records == null) {
                records = Collections.emptyList();
            } else {
                records = Collections.unmodifiableList(new ArrayList<WeaponStockRecord>(records));
            }
            result.put(category, records);
        }
        return Collections.unmodifiableMap(result);
    }

    private static Map<StockItemType, Map<StockCategory, List<WeaponStockRecord>>> immutableTypeCategoryMap(Map<StockCategory, List<WeaponStockRecord>> recordsByCategory) {
        Map<StockItemType, Map<StockCategory, List<WeaponStockRecord>>> result =
                new EnumMap<StockItemType, Map<StockCategory, List<WeaponStockRecord>>>(StockItemType.class);
        for (StockItemType itemType : StockItemType.values()) {
            Map<StockCategory, List<WeaponStockRecord>> grouped =
                    new EnumMap<StockCategory, List<WeaponStockRecord>>(StockCategory.class);
            for (StockCategory category : StockCategory.values()) {
                List<WeaponStockRecord> source = recordsByCategory.get(category);
                List<WeaponStockRecord> records = new ArrayList<WeaponStockRecord>();
                if (source != null) {
                    for (int i = 0; i < source.size(); i++) {
                        WeaponStockRecord record = source.get(i);
                        if (record != null && itemType.equals(record.getItemType())) {
                            records.add(record);
                        }
                    }
                }
                grouped.put(category, Collections.unmodifiableList(records));
            }
            result.put(itemType, Collections.unmodifiableMap(grouped));
        }
        return Collections.unmodifiableMap(result);
    }

    private static Map<String, WeaponStockRecord> immutableWeaponMap(Map<StockCategory, List<WeaponStockRecord>> recordsByCategory) {
        Map<String, WeaponStockRecord> result = new java.util.HashMap<String, WeaponStockRecord>();
        for (StockCategory category : StockCategory.values()) {
            List<WeaponStockRecord> records = recordsByCategory.get(category);
            if (records == null) {
                continue;
            }
            for (int i = 0; i < records.size(); i++) {
                WeaponStockRecord record = records.get(i);
                result.put(record.getWeaponId(), record);
            }
        }
        return Collections.unmodifiableMap(result);
    }

    private static List<WeaponStockRecord> immutableAllRecords(Map<StockCategory, List<WeaponStockRecord>> recordsByCategory) {
        List<WeaponStockRecord> result = new ArrayList<WeaponStockRecord>();
        for (StockCategory category : StockCategory.values()) {
            List<WeaponStockRecord> records = recordsByCategory.get(category);
            if (records != null) {
                result.addAll(records);
            }
        }
        return Collections.unmodifiableList(result);
    }
}

File: src/main/kotlin/weaponsprocurement/core/WeaponStockSnapshot.kt
package weaponsprocurement.core

import com.fs.starfarer.api.campaign.econ.MarketAPI
import java.util.ArrayList
import java.util.Collections
import java.util.EnumMap
import java.util.HashMap

class WeaponStockSnapshot(
    private val market: MarketAPI?,
    private val ownedSourcePolicy: OwnedSourcePolicy,
    private val sortMode: StockSortMode,
    private val includeBlackMarket: Boolean,
    sourceMode: StockSourceMode?,
    recordsByCategory: Map<StockCategory, List<WeaponStockRecord>>,
) {
    private val sourceMode: StockSourceMode = sourceMode ?: StockSourceMode.LOCAL
    private val recordsByCategory: Map<StockCategory, List<WeaponStockRecord>> =
        immutableCategoryMap(recordsByCategory)
    private val recordsByTypeAndCategory: Map<StockItemType, Map<StockCategory, List<WeaponStockRecord>>> =
        immutableTypeCategoryMap(this.recordsByCategory)
    private val recordsByItemKey: Map<String, WeaponStockRecord> =
        immutableItemKeyMap(this.recordsByCategory)
    private val allRecords: List<WeaponStockRecord> = immutableAllRecords(this.recordsByCategory)
    private val totalRecords: Int = allRecords.size

    fun getMarket(): MarketAPI? = market

    fun getOwnedSourcePolicy(): OwnedSourcePolicy = ownedSourcePolicy

    fun getSortMode(): StockSortMode = sortMode

    fun isIncludeBlackMarket(): Boolean = includeBlackMarket

    fun getSourceMode(): StockSourceMode = sourceMode

    fun getRecords(category: StockCategory?): List<WeaponStockRecord> {
        return recordsByCategory[category] ?: Collections.emptyList()
    }

    fun getRecords(itemType: StockItemType?, category: StockCategory?): List<WeaponStockRecord> {
        val byCategory = recordsByTypeAndCategory[itemType] ?: return Collections.emptyList()
        return byCategory[category] ?: Collections.emptyList()
    }

    fun getCount(itemType: StockItemType?): Int {
        var count = 0
        val byCategory = recordsByTypeAndCategory[itemType] ?: return 0
        for (category in StockCategory.values()) {
            count += byCategory[category]?.size ?: 0
        }
        return count
    }

    fun getCount(category: StockCategory?): Int = getRecords(category).size

    fun getTotalRecords(): Int = totalRecords

    fun getRecord(itemKey: String?): WeaponStockRecord? {
        if (itemKey == null) return null
        return recordsByItemKey[itemKey]
    }

    fun getAllRecords(): List<WeaponStockRecord> = allRecords

    fun getMarketName(): String {
        if (StockSourceMode.FIXERS == sourceMode) return GlobalWeaponMarketService.FIXERS_MARKET_NAME
        if (StockSourceMode.SECTOR == sourceMode) return GlobalWeaponMarketService.SECTOR_MARKET_NAME
        return market?.name ?: "No market context"
    }

    companion object {
        private fun immutableCategoryMap(
            source: Map<StockCategory, List<WeaponStockRecord>>,
        ): Map<StockCategory, List<WeaponStockRecord>> {
            val result = EnumMap<StockCategory, List<WeaponStockRecord>>(StockCategory::class.java)
            for (category in StockCategory.values()) {
                val records = source[category]
                result[category] = if (records == null) {
                    Collections.emptyList()
                } else {
                    Collections.unmodifiableList(ArrayList(records))
                }
            }
            return Collections.unmodifiableMap(result)
        }

        private fun immutableTypeCategoryMap(
            recordsByCategory: Map<StockCategory, List<WeaponStockRecord>>,
        ): Map<StockItemType, Map<StockCategory, List<WeaponStockRecord>>> {
            val result = EnumMap<StockItemType, Map<StockCategory, List<WeaponStockRecord>>>(StockItemType::class.java)
            for (itemType in StockItemType.values()) {
                val grouped = EnumMap<StockCategory, List<WeaponStockRecord>>(StockCategory::class.java)
                for (category in StockCategory.values()) {
                    val source = recordsByCategory[category]
                    val records = ArrayList<WeaponStockRecord>()
                    if (source != null) {
                        for (record in source) {
                            if (itemType == record.itemType) records.add(record)
                        }
                    }
                    grouped[category] = Collections.unmodifiableList(records)
                }
                result[itemType] = Collections.unmodifiableMap(grouped)
            }
            return Collections.unmodifiableMap(result)
        }

        private fun immutableItemKeyMap(
            recordsByCategory: Map<StockCategory, List<WeaponStockRecord>>,
        ): Map<String, WeaponStockRecord> {
            val result = HashMap<String, WeaponStockRecord>()
            for (category in StockCategory.values()) {
                val records = recordsByCategory[category] ?: continue
                for (record in records) {
                    result[record.itemKey] = record
                }
            }
            return Collections.unmodifiableMap(result)
        }

        private fun immutableAllRecords(
            recordsByCategory: Map<StockCategory, List<WeaponStockRecord>>,
        ): List<WeaponStockRecord> {
            val result = ArrayList<WeaponStockRecord>()
            for (category in StockCategory.values()) {
                val records = recordsByCategory[category]
                if (records != null) result.addAll(records)
            }
            return Collections.unmodifiableList(result)
        }
    }
}
